runtime/core/http/http_server.cc

```diff
@@ -0,0 +1,206 @@
+                                          *decode_config_);
+  // Start decoder thread
+  decode_thread_ =
+      std::make_shared<std::thread>(&ConnectionHandler::DecodeThreadFunc, this);
+}
+
+void ConnectionHandler::OnSpeechEnd() {
+  LOG(INFO) << "Received speech end signal";
+  if (feature_pipeline_ != nullptr) {
+    feature_pipeline_->set_input_finished();
+  }
+}
+
+void ConnectionHandler::OnFinalResult(const std::string& result) {
+  LOG(INFO) << "Final result: " << result;
+  json::value rv = {
+      {"status", "ok"}, {"type", "final_result"}, {"nbest", result}};
+  std::string message = json::serialize(rv);
+  res_.get()->body() = message;
+  http::write(socket_, *res_.get(), ec_);
+}
+
+void ConnectionHandler::OnSpeechData(const std::string& message) {
+  std::size_t decode_size =
+      beast::detail::base64::decoded_size(message.length());
+  int num_samples = decode_size / sizeof(int16_t);
+  int16_t decode_data[num_samples];
+  beast::detail::base64::decode(decode_data, message.c_str(), message.length());
+
+  // Read binary PCM data
+  VLOG(2) << "Received " << num_samples << " samples";
+  CHECK(feature_pipeline_ != nullptr);
+  CHECK(decoder_ != nullptr);
+  feature_pipeline_->AcceptWaveform(decode_data, num_samples);
+}
+
+std::string ConnectionHandler::SerializeResult(bool finish) {
+  json::array nbest;
+  for (const DecodeResult& path : decoder_->result()) {
+    json::object jpath({{"sentence", path.sentence}});
+    if (finish) {
+      json::array word_pieces;
+      for (const WordPiece& word_piece : path.word_pieces) {
+        json::object jword_piece({{"word", word_piece.word},
+                                  {"start", word_piece.start},
+                                  {"end", word_piece.end}});
+        word_pieces.emplace_back(jword_piece);
+      }
+      jpath.emplace("word_pieces", word_pieces);
+    }
+    nbest.emplace_back(jpath);
+
+    if (nbest.size() == nbest_) {
+      break;
+    }
+  }
+  return json::serialize(nbest);
+}
+
+void ConnectionHandler::DecodeThreadFunc() {
+  try {
+    while (true) {
+      DecodeState state = decoder_->Decode();
+      if (state == DecodeState::kEndFeats || state == DecodeState::kEndpoint) {
+        decoder_->Rescoring();
+        std::string result = SerializeResult(true);
+        OnFinalResult(result);
+        break;
+      }
+    }
+  } catch (std::exception const& e) {
+    LOG(ERROR) << e.what();
+  }
+}
+
+void ConnectionHandler::OnError(const std::string& message) {
+  json::value rv = {{"status", "failed"}, {"message", message}};
+  res_.get()->body() = json::serialize(rv);
+  http::write(socket_, *res_.get(), ec_);
+  // Send a TCP shutdown
+  socket_.shutdown(tcp::socket::shutdown_send, ec_);
+}
+
+void ConnectionHandler::OnText(const std::string& message) {
+  LOG(INFO) << message;
+  json::value v = json::parse(message);
+  if (v.is_object()) {
+    json::object obj = v.get_object();
+    if (obj.find("signal") != obj.end()) {
```

Comment:
singal start and end are not required here.

Reply:
deleted signal start and end message